plasmapy_nei/nei/nei.py

```diff
@@ -1,7 +1,1363 @@
+                initial_states.ionic_fractions.keys()
+            )  # TODO IonizationStates
+        elif initial_states is None:
+            self._ionstates = None
+        else:
+            raise TypeError("Expecting an IonizationStates instance.")
+
+    @property
+    def results(self) -> SimulationResults:
+        """
+        Return the `~plasmapy_nei.nei.SimulationResults` class instance that
+        corresponds to the simulation results.
+
+        """
+        if self._results is not None:
+            return self._results
+        else:
+            raise AttributeError("The simulation has not yet been performed.")
+
+    @property
+    def final(self) -> IonizationStates:
+        """
+        Return the ionization states of the plasma at the end of the
+        simulation.
+        """
+        try:
+            return self._final
+        except AttributeError:
+            raise NEIError("The simulation has not yet been performed.") from None
+
+    def _initialize_simulation(self):
+
+        self._results = SimulationResults(
+            initial=self.initial,
+            n_init=self.hydrogen_number_density(self.time_start),
+            T_e_init=self.electron_temperature(self.time_start),
+            max_steps=self.max_steps,
+            time_start=self.time_start,
+        )
+        self._old_time = self.time_start.to(u.s)
+        self._new_time = self.time_start.to(u.s)
+
+    def simulate(self) -> SimulationResults:
+        """
+        Perform a non-equilibrium ionization simulation.
+
+        Returns
+        -------
+        results: `~plasmapy_nei.classes.Simulation`
+            The results from the simulation (which are also stored in
+            the ``results`` attribute of the `~plasmapy_nei.nei.NEI`
+            instance this method was called from.
+
+        """
+
+        self._initialize_simulation()
+
+        for step in range(self.max_steps):
+            try:
+                self.set_timestep()
+                self.time_advance()
+            except StopIteration:
+                break
+            except Exception as exc:
+                raise NEIError(f"Unable to complete simulation.") from exc
+
+        self._finalize_simulation()
+
+        # Is there a way to use the inspect package or something similar
+        # to only return self.results if it is in an expression where
+
+        return self.results
+
+    def _finalize_simulation(self):
+        self._results._cleanup()
+
+        final_ionfracs = {
+            element: self.results.ionic_fractions[element][-1, :]
+            for element in self.elements
+        }
+
+        self._final = IonizationStates(
+            inputs=final_ionfracs,
+            abundances=self.abundances,
+            n=np.sum(self.results.number_densities["H"][-1, :]),  # modify this later?,
+            T_e=self.results.T_e[-1],
+            tol=1e-6,
+        )
+
+        if not np.isclose(self.time_max / u.s, self.results.time[-1] / u.s):
```

Comment:
Since you're policing units, why not just do...

```suggestion
        if not np.isclose(self.time_max.value, self.results.time[-1].vlaue):
```